src/Psalm/Internal/Analyzer/StatementsAnalyzer.php

```diff
@@ -769,12 +775,32 @@ public function checkUnreferencedVars(array $stmts): void
                 && $this->data_flow_graph instanceof VariableUseGraph
                 && !$this->data_flow_graph->isVariableUsed($assignment_node)
             ) {
-                $issue = new UnusedVariable(
-                    $var_id . ' is never referenced or the value is not used',
-                    $original_location
+                $original_foreach_stmt = NodeFinder::find(
```

Comment:
My worry here is that this is a bunch of unnecessary work for codebases that have grandfathered-in issues on each Psalm run.

What if the `$statements_analyzer` (created afresh for each function analysis) held a list of foreach variable positions? That way the comparison wouldn't have to loop recurse the AST.

Reply:
Yeah, that would work and is actually easy to implement by storing `$stmt->valueVar` locations in `ForeachAnalyzer::analyze()` into `$statements_analyzer`.